### PR TITLE
Added code for Dynamic Programming Algorithms

### PR DESCRIPTION
**Pull request by abhijeet147**
`Please fill all the details and checkbox, so we can merge PR faster.`

### Issue no.(must) - 946

Pr will be closed and marked as spam. If issue number not found or issue was assigned to someone else. 
Marking as spam can block your account from HacktoberFest.
### Self Check(Tick After Making pull Request)

- [x] This issue was assigned to me if not it will marked as spam.
- [x] My file is in proper folder
- [x] I am following clean code and Documentation.My code is well linted and I am using flake8 standard for python
- [x] I have added title and what program will do first in the file

## acknowledgment
- [x] I have tested the script and distributed it under an MIT license.

### README - [How to Contribute](https://github.com/Py-Contributors/AlgorithmsAndDataStructure/blob/master/CONTRIBUTING.md)
[SpiralMatrixzip.zip](https://github.com/Py-Contributors/AlgorithmsAndDataStructure/files/12805191/SpiralMatrixzip.zip)
[Schedulung_algorithmzip.zip](https://github.com/Py-Contributors/AlgorithmsAndDataStructure/files/12805194/Schedulung_algorithmzip.zip)
[Pattern_searchingzip.zip](https://github.com/Py-Contributors/AlgorithmsAndDataStructure/files/12805195/Pattern_searchingzip.zip)
[Pathfinding_algorithmzip.zip](https://github.com/Py-Contributors/AlgorithmsAndDataStructure/files/12805197/Pathfinding_algorithmzip.zip)

### CODE REVIEW

File: SpiralMatrix.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+using namespace std;
+
+void spiral_matrix() {
+    int i = 0;  // starting row index
+    int j = 0;  // starting column index
+    int c = 0;
+    int no;
+
+    cout << "Please enter number of rows of the spiral matrix: ";
+    cin >> no;
```
Comment (cankush625) [inline SpiralMatrix.cpp:13]: This should be accepted as a function parameter.

File: SpiralMatrix.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
+#include <iostream>
+#include <vector>
+
+using namespace std;
+
+void spiral_matrix() {
+    int i = 0;  // starting row index
+    int j = 0;  // starting column index
+    int c = 0;
+    int no;
```
Comment (cankush625) [inline SpiralMatrix.cpp:10]: Variables names can be meaningful.

File: Duplicate_subtree.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+std::string inorder(Node* root, std::unordered_map<std::string, int>& m) {
+    if (root == nullptr) {
+        return "";
+    }
+
+    std::string s = "(";
+    s += inorder(root->left, m);
+    s += std::to_string(root->data);
+    s += inorder(root->right, m);
+    s += ")";
```
Comment (cankush625) [inline Duplicate_subtree.cpp:23]: Variable name can be better

File: Duplicate_subtree.cpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+class Node {
+public:
+    int data;
+    Node* left;
+    Node* right;
+
+    Node(int data) : data(data), left(nullptr), right(nullptr) {}
+};
+
+std::string inorder(Node* root, std::unordered_map<std::string, int>& m) {
```
Comment (cankush625) [inline Duplicate_subtree.cpp:14]: Parameter name for `m` can be better

**Review by cankush625 — CHANGES_REQUESTED** [review]
Added comments for code quality. Not sure about the execution of these programs.

**Comment by codeperfectplus** [comment]
Please make the @abhijeet147 changes as suggested by @cankush625.
We are focusing on better code quality now 
Thanks

**Comment by codeperfectplus** [comment]
Don't add Zip files.

**Review by Mr-vabs — APPROVED** [review]
Seems good to go

**Pull request closed without merging**